src/Elastic.Apm/Agent.cs

```diff
@@ -162,16 +168,32 @@ private static bool CheckAndAddFilter(Action<PayloadSenderV2> action)
 
 		public static void Setup(AgentComponents agentComponents)
 		{
-			if (LazyApmAgent.IsValueCreated)
-				agentComponents?.Logger?.Error()
-					?.Log("The singleton APM agent has" +
-						" already been instantiated and can no longer be configured. Reusing existing instance");
-
-			agentComponents?.Logger?.Trace()
-				?.Log("Initialization - Agent.Setup called. Callstack: {callstack}", new StackTrace().ToString());
-
-			Components = agentComponents;
-			_isConfigured = true;
+			lock (InitializationLock)
+			{
+				if (LazyApmAgent.IsValueCreated)
+				{
+					Components?.Logger?.Error()
+						?.Log("The singleton APM agent has" +
+							" already been instantiated and can no longer be configured. Reusing existing instance. "
+							+ "Callstack: {callstack}", new StackTrace().ToString());
+
+					// Above line logs on the already configured `Components`
+					// In order to let the caller know, we also log on the logger of the rejected `agentComponents`
+					agentComponents?.Logger?.Error()
+						?.Log("The singleton APM agent has" +
+							" already been instantiated and can no longer be configured. Reusing existing instance. "
+							+ "Callstack: {callstack}", new StackTrace().ToString());
+
+					return;
+				}
+
+				agentComponents?.Logger?.Trace()
+					?.Log("Initialization - Agent.Setup called");
```

Comment:
We don't print the callstack here, because below we force initialization with `LazyApmAgent.Value` and that getter prints the callstack.